### PR TITLE
migrated index and shard components away from guice

### PR DESCRIPTION
**Pull request by dobe**
there is now no dependency anymore on shard or index
level injectors

### CODE REVIEW

File: blob/src/main/java/io/crate/blob/v2/BlobIndicesService.java [hunk 1]
```diff
@@ -74,7 +72,7 @@ public BlobIndicesService(Settings settings,
         super(settings);
         this.clusterService = clusterService;
         this.indicesLifecycle = indicesLifecycle;
-        this.listener = new LifecycleListener();
+        indicesLifecycle.addListener(new LifecycleListener());
```
Comment (mfussenegger) [inline blob/src/main/java/io/crate/blob/v2/BlobIndicesService.java]: We've had issues before using `addListener` in the constructor (see 4f68d8b4c6e87a051e288ee437c9f6f3700a8c9f). Isn't that a problem here as well?

File: sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java [hunk 2]
```diff
@@ -349,32 +411,34 @@ private CrateCollector getShardsCollector(RoutedCollectPhase collectPhase,
...
+                    ShardId shardId = new ShardId(indexName, shard);
+                    ShardCollectorProvider shardCollectorProvider = getCollectorProviderSafe(shardId);
+                    Object[] row = shardCollectorProvider.getRowForShard(normalizedPhase);
                     if (row != null) {
                         rows.add(row);
                     }
                 } catch (ShardNotFoundException | IllegalIndexShardStateException e) {
                     unassignedShards.add(toUnassignedShard(new ShardId(indexName, shard)));
                 } catch (Throwable t) {
+                    t.printStackTrace();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java]: remove

File: sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java [hunk 3]
```diff
@@ -349,32 +411,34 @@ private CrateCollector getShardsCollector(RoutedCollectPhase collectPhase,
...
             }
         }
         if (!unassignedShards.isEmpty()) {
             // since unassigned shards aren't really on any node we use the collectPhase which is NOT normalized here
             // because otherwise if _node was also selected it would contain something which is wrong
-            for (Object[] objects : Iterables.transform(
-                systemCollectSource.toRowsIterable(collectPhase, unassignedShards, false), Row.MATERIALIZE)) {
+            for (Object[] objects : StreamSupport.stream(
+                systemCollectSource.toRowsIterable(collectPhase, unassignedShards, false).spliterator(), false)
+                .map(Row.MATERIALIZE::apply).collect(Collectors.toList())) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java]: I think `.collect(Collectors.ToList)` isn't really necessary here.  `rows.add(objects)` could be done with a `forEach` on the stream directly

File: sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java [hunk 4]
```diff
@@ -349,32 +411,34 @@ private CrateCollector getShardsCollector(RoutedCollectPhase collectPhase,
...
                 rows.add(objects);
             }
         }
 
         if (collectPhase.orderBy() != null) {
-            Collections.sort(rows, OrderingByPosition.arrayOrdering(collectPhase).reverse());
+            rows.sort(OrderingByPosition.arrayOrdering(collectPhase).reverse());
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java]: just curious - is this faster or why the change here?

File: sql/src/main/java/io/crate/operation/reference/sys/shard/ShardRecoveryExpression.java [hunk 5]
```diff
@@ -37,32 +36,31 @@
     private static final String SIZE = "size";
     private static final String FILES = "files";
 
-
     public ShardRecoveryExpression(IndexShard indexShard) {
-        addChildImplementations(indexShard.recoveryState());
+        addChildImplementations(indexShard);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/reference/sys/shard/ShardRecoveryExpression.java]: maybe add a comment somewhere that recoveryState is mutable and that we want to get the latest state at runtime?

File: blob/src/main/java/io/crate/blob/v2/BlobIndicesService.java (hunk 1 above)
Comment (dobe) [inline blob/src/main/java/io/crate/blob/v2/BlobIndicesService.java]: should not, this was guice related and there is a special test for it

File: sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java (hunk 4 above)
Comment (dobe) [inline sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java]: hm, actually it was suggested by idea, however i think ther is no perf diff, but i think it is more OO without that static method

File: sql/src/main/java/io/crate/operation/reference/sys/shard/ShardRecoveryExpression.java (hunk 5 above)
Comment (dobe) [inline sql/src/main/java/io/crate/operation/reference/sys/shard/ShardRecoveryExpression.java]: added a note in ShardRecoveryStateExpression
